Update guide to use JSON-B instead of Jackson

KyleAure opened the pull request:
Will not work until Java EE8 is released. Updating guide using JSON-B instead of Jackson.  Also using the newest version of JSON-P.

Kubik42 commented:
Hi @KyleAure. Thanks for the PR! I'll take a look at it sometime this week.

Kubik42 commented on README.adoc:
```diff
@@ -112,34 +112,41 @@ include::finish/src/main/java/io/openliberty/guides/consumingrest/model/Artist.j
 does not require this annotation since all instance members match the JSON keys by name.
 
-* The `@JsonIgnore` annotation is used to define an object property that does not map to any JSON
+* The `@JsonbCreator` and `@JsonProperty` annotations are used to annotate a custom constructor.
```
Shouldn't this be `JsonbProperty`?

Kubik42 commented on README.adoc:
```diff
@@ -112,34 +112,41 @@ include::finish/src/main/java/io/openliberty/guides/consumingrest/model/Artist.j
+* The `@JsonbCreator` and `@JsonProperty` annotations are used to annotate a custom constructor.
+This is required for proper parameter substitution when a custom constructor is used.
+
+* The `@JsonTransient` annotation is used to define an object property that does not map to any JSON
```
Same as above comment, shouldn't this be `JsonbTransient`?

Kubik42 commented on README.adoc:
```diff
@@ -148,83 +155,84 @@ Write a simple `Consumer` class that consumes the artist JSON using both JSON-P
+which has been done for you:
 
-[source, java, indent=0]
+[source, xml]
```
Can you add `role="no_copy` to distinguish this listing block since it doesn't need to be copied.

Kubik42 commented on README.adoc:
```diff
@@ -148,83 +155,84 @@ Write a simple `Consumer` class that consumes the artist JSON using both JSON-P
+To bind the JSON into an `Artist` array, use the `Artist[]` entity type in the `readEntity` call:
 
-[source, xml]
+[source, java, indent = 0]
```
Same `role="no_copy"` bit

Kubik42 commented on README.adoc:
```diff
@@ -148,83 +155,84 @@ Write a simple `Consumer` class that consumes the artist JSON using both JSON-P
 [source, java, indent=0]
 ----
-include::finish/src/main/java/io/openliberty/guides/consumingrest/Consumer.java[tags=consumeWithJackson]
+include::finish/src/main/java/io/openliberty/guides/consumingrest/service/ArtistResource.java[tags=getJsonString]
```
Can you convert the tabs to spaces for this method? Otherwise it renders with too much left space.

Kubik42 commented on README.adoc:
```diff
@@ -249,26 +257,31 @@ You can find more detailed explanations on how to define your own system propert
 [source, java, indent = 0]
 ----
-include::finish/src/test/java/it/io/openliberty/guides/consumingrest/ConsumingRestTest.java[tags=assertResponse]
+include::finish/src/test/java/it/io/openliberty/guides/consumingrest/ConsumingRestTest.java[tags=testArtistDeserialization]
```
Fix indentation in this method as well.

KyleAure commented:
@Kubik42 Thank you for the feedback!  I have made the changes you have requested.

yeekangc commented:
This should not be merged until 18.0.0.2 is released.  @KyleAure @Kubik42 

We certainly should get ready and test with 18.0.0.2 daily builds.

sishida commented on README.adoc:
```diff
@@ -19,20 +19,20 @@
-list of artists and their albums using two different approaches. The first approach consists of using
-the JSON-P API to process the JSON. The second approach focuses on using a JSON binding library like
-Jackson to convert JSON messages into Java objects directly.
+You will learn how to access a REST service, serialize a java object containing a
```
- capitalize "Java"
- say "that contains" instead of "containing"

sishida commented on README.adoc:
```diff
@@ -19,20 +19,20 @@
-the JSON-P API to process the JSON. The second approach focuses on using a JSON binding library like
-Jackson to convert JSON messages into Java objects directly.
+You will learn how to access a REST service, serialize a java object containing a
+list of artists and their albums, and deserialize the returned JSON resources using
```
What about, "...and use two different approaches to deserialize the returned JSON resources."?

sishida commented on README.adoc:
```diff
@@ -19,20 +19,20 @@
+list of artists and their albums, and deserialize the returned JSON resources using
+two different approaches. The first approach consists of using JSON-B (binding) API
+to convert JSON messages into Java objects directly.  The second approach consists of
+using the JSON-P (JSON with Padding) API to process the JSON.
```
- I'd say, "**the** JSON-B..."
- We try to avoid parenthesis. Could we either remove the content in parenthesis or somehow rewrite these sentences to include the content but remove the parenthesis?
- Maybe also move the adverb like so: "...convert JSON messages directly into Java objects."
- Remove the extra space between the sentences on line 31.

sishida commented on README.adoc:
```diff
@@ -112,34 +112,41 @@ include::finish/src/main/java/io/openliberty/guides/consumingrest/model/Artist.j
 
-* Cannot have a constructor unless it is empty.
-* Must contain the same number of instance members as there are keys in the target JSON.
+JSON-B is a new feature introduced with Java EE 8 and strengthens Java's support for JSON.
```
- Let's omit the word "new." You already say that it's introduced with Java EE 8, and that statement should suffice. "New" can be such a relative term and might feel outdated if this guide stays around a while.
- I also suggest, "Java support" instead of "Java's support."

sishida commented on README.adoc:
```diff
@@ -112,34 +112,41 @@ include::finish/src/main/java/io/openliberty/guides/consumingrest/model/Artist.j
-* Cannot have a constructor unless it is empty.
-* Must contain the same number of instance members as there are keys in the target JSON.
+JSON-B is a new feature introduced with Java EE 8 and strengthens Java's support for JSON.
+This guide previously used Jackson for JSON binding.  JSON-B requires, by default, a
```
- I'd remove the sentence, "This guide previously used Jackson for JSON binding."
- Place "By default" at the beginning of the sentence. "By default, JSON-B requires..."

sishida commented on README.adoc:
```diff
@@ -112,34 +112,41 @@ include::finish/src/main/java/io/openliberty/guides/consumingrest/model/Artist.j
-* Must contain the same number of instance members as there are keys in the target JSON.
+JSON-B is a new feature introduced with Java EE 8 and strengthens Java's support for JSON.
+This guide previously used Jackson for JSON binding.  JSON-B requires, by default, a
+public default no-arguments constructor for deserialization.
```
I've seen "no-argument" instead. Grammatically, the singular "argument" makes more sense to me, but please go with whatever is technically correct. ;)

sishida commented on README.adoc:
```diff
@@ -112,34 +112,41 @@ include::finish/src/main/java/io/openliberty/guides/consumingrest/model/Artist.j
+public default no-arguments constructor for deserialization.
 
-While Jackson annotations are not always necessary, they are considered good practice when used:
+The JSON-B engine includes a set of default mapping rules which can be run without
```
Add a comma before "which."

sishida commented on README.adoc:
```diff
@@ -112,34 +112,41 @@ include::finish/src/main/java/io/openliberty/guides/consumingrest/model/Artist.j
 
-While Jackson annotations are not always necessary, they are considered good practice when used:
+The JSON-B engine includes a set of default mapping rules which can be run without
+any customization annotations or custom configuration. There are instances where you may
```
How about, "In some instances, you might find it useful to deserialize a JSON message..."
- We try to avoid "there" as the sentence subject because it doesn't really add info on its own.
- We also use "might" instead of "may" to indicate possibility.

sishida commented on README.adoc:
```diff
@@ -112,34 +112,41 @@ include::finish/src/main/java/io/openliberty/guides/consumingrest/model/Artist.j
+The JSON-B engine includes a set of default mapping rules which can be run without
+any customization annotations or custom configuration. There are instances where you may
+find it useful to deserialize a JSON message with only certain properties, specific
+property names, or have a class with a custom constructor. In these cases,
```
When writing a list, try to make the items in the list parallel, such as all nouns, all verbs, etc. In this list, "have a class with a custom constructor" is the odd one out because all the other items in the list are nouns, but this last one starts with a verb, "have," instead. Could you change it by removing the "have" and use "classes with custom constructors"?

sishida commented on README.adoc:
```diff
@@ -112,34 +112,41 @@ include::finish/src/main/java/io/openliberty/guides/consumingrest/model/Artist.j
 
-* The `@JsonIgnore` annotation is used to define an object property that does not map to any JSON
+* The `@JsonbCreator` and `@JsonbProperty` annotations are used to annotate a custom constructor.
+This is required for proper parameter substitution when a custom constructor is used.
```
- What about, "You can use the `@JsonbCreator` and `@JsonbProperty` annotations to annotate a custom constructor." These edits remove passive voice and change it to active voice.
- Avoid using "this" as a standalone word but instead say, "this _thing_." "This" by itself can be ambiguous. Maybe say, "The annotations are required for proper...'

sishida commented on README.adoc:
```diff
@@ -112,34 +112,41 @@ include::finish/src/main/java/io/openliberty/guides/consumingrest/model/Artist.j
 property. While the use of this annotation is good practice, it is only necessary for serialization.
 
+For more information on customization with JSON-B please see the
+http://json-b.net/users-guide.html[JSON Binding 1.0 Users Guide].
```
- Add a comma after "JSON-B"
- Remove "please." 
- I noticed that the title of the "Users Guide" shows "Json" with the lower case "son." The text has the letters as capital, though, and we have them as capital, so I suppose it's not a big deal either way.

sishida commented on README.adoc:
```diff
@@ -148,83 +155,84 @@ Write a simple `Consumer` class that consumes the artist JSON using both JSON-P
 
-Write the `collectArtists` and `collectAlbums` helpers. These methods will parse that JSON and
-collect its objects into individual POJOs:
+To include the JSON-B provider in your project, add the following dependency to your `pom.xml`,
```
Specify "`pom.xml` file", which can be useful if we ever translate the guides.

sishida commented on README.adoc:
```diff
@@ -148,83 +155,84 @@ Write a simple `Consumer` class that consumes the artist JSON using both JSON-P
-Write the `collectArtists` and `collectAlbums` helpers. These methods will parse that JSON and
-collect its objects into individual POJOs:
+To include the JSON-B provider in your project, add the following dependency to your `pom.xml`,
+which has been done for you:
```
How about, "which is already done for you" instead?

sishida commented on README.adoc:
```diff
@@ -148,83 +155,84 @@ Write a simple `Consumer` class that consumes the artist JSON using both JSON-P
-approach to return the total number of artists present in the JSON.
-
-`start/src/main/java/io/openliberty/guides/consumingrest/service/ArtistResource.java`:
+Write the method for consuming the REST service using JSON-B in the `Consumer` class:
```
What about, "Use JSON-B in the `Consumer` class to write the method for consuming the REST service:"?

sishida commented:
I am still working through the edits for the latest changes. I will continue tomorrow if I can't finish today.

sishida commented on README.adoc:
```diff
@@ -148,83 +155,84 @@ Write a simple `Consumer` class that consumes the artist JSON using both JSON-P
 
-Write the method for consuming the REST service using Jackson's JSON binding in the `Consumer` class:
+Write the `getJsonString` and `getTotalAlbums` methods in the `ArtistResource` class. The `getTotalAlbums`
+method will use JSON-B to return the total number of albums present in the JSON for a particular artist
```
- Say, "The `getTotalAlbums` method **uses**" instead of "will use." We try to keep the sentences in present tense as much as possible.
- I think we can end the sentence at "...for particular artist." Then, we can make a new sentence, "JSON-B returns -1 if this artist does not exist."
- Remove the extra space after "JSON-B" in line 192.
- Try to avoid first person pronouns like "I" and "we" (line 193). Perhaps say, "The `getJsonString` method returns the JSON as a string that you can use later for testing."

sishida commented on README.adoc:
```diff
@@ -148,83 +155,84 @@ Write a simple `Consumer` class that consumes the artist JSON using both JSON-P
 ----
 
-To bind the JSON into an `Artist` array, use the `Artist[]` entity type in the `readEntity` call:
+Write the `collectArtists` and `collectAlbums` helpers. These methods will parse that JSON and
```
"These methods parse the JSON and..."

sishida commented on README.adoc:
```diff
@@ -148,83 +155,84 @@ Write a simple `Consumer` class that consumes the artist JSON using both JSON-P
 
-To bind the JSON into an `Artist` array, use the `Artist[]` entity type in the `readEntity` call:
+Write the `collectArtists` and `collectAlbums` helpers. These methods will parse that JSON and
+collect its objects into individual POJOs.  Notice that we are able to use our custom constructors
```
- Remove extra space between sentences on line 213.
- Remove first person pronoun. I'd rewrite as, "You can use the custom constructors to..."

sishida commented on README.adoc:
```diff
@@ -148,83 +155,84 @@ Write a simple `Consumer` class that consumes the artist JSON using both JSON-P
+It is worth noting that in this project the methods that we created in the `Consumer`
+class could be written directly in the `ArtistResource` class.  However, if you are
+consuming a REST resource from a third party service it is best practice to separate
+your GET and PULL requests from your data consumption.
```
Suggested rewrite:
"The methods that you created in the `Consumer` class can be written directly in the `ArtistResource` class. However, if you are consuming a REST resource from a third party service, you can separate your GET and PULL requests from your data consumption."

sishida commented on README.adoc:
```diff
@@ -249,26 +257,31 @@ You can find more detailed explanations on how to define your own system propert
-and that the correct number of artist is returned.
+=== Testing the binding process
+In order for your test classes to have access to JSON-B you must add the following dependency to your `pom.xml`,
+which has been done for you:
```
"In order" doesn't really add to the sentence and can simply be omitted. 

How about something like, "For your test classes to have access to JSON-B, add the following dependency to your `pom.xml` file, which is already done for you:"

KyleAure commented:
@sishida I went through and made the changes you suggested.  Thank you.

sishida commented on README.adoc:
```diff
@@ -249,26 +254,31 @@ You can find more detailed explanations on how to define your own system propert
 
-Write the `assertResponse` helper method that asserts that the response code you receive is valid (200):
+Write the `testArtistDeserialization` test case.  This test checks that Artist instances created from
+the REST data and those that are hardcoded perform the same.
```
- Remove extra space on line 271.
- Earlier, I saw "Artist" formatted as `Artist`. Should we use the code phrase format on line 271 also?

sishida commented on README.adoc:
```diff
@@ -307,17 +334,36 @@ output:
+
 ----
 
+== Conclusions
```
We generally don't include any kind of conclusion section in our guides. If this information is necessary to include in the guide, could you place it earlier in the file in another section? Perhaps you could create a new section with its own title and put the new section around the sections on "POJOs" and "Using JSON-B" at the beginning of the guide.

sishida commented on README.adoc:
```diff
@@ -307,17 +334,36 @@ output:
+and allows for more customization for serializing and deserializing.  We would recommend,
+if you have not already, to try using JSON-B for your next project.  If you want
+more information on JSON-B and how to serialize POJOs into JSON strings please refer to the
+http://json-b.net/users-guide.html[JSON Binding 1.0 Users Guide].
```
Include only one space between sentences.

sishida replied:
Maybe under "What you'll learn" you could include a "Differences between JSON-B and JSON-P" section.

Suggested changes:

JSON-B allows you to directly serialize and deserialize POJOs. This API gives you options for working with JSON resources.

In contrast, JSON-P requires you to use helper methods to process the JSON response. This tactic is more straightforward, but it can be cumbersome with more complex classes.

JSON-B is built on top of the existing JSON-P API. JSON-B can do everything that JSON-P can and allows for more customization for serializing and deserializing. For more information on JSON-B and how to serialize POJOs into JSON strings, see the http://json-b.net/users-guide.html[JSON Binding 1.0 Users Guide].

KyleAure commented:
Thank you again @sishida for all your help with this guide!  I made the changes you suggested and removed the conclusion and moved it more towards the top.  The build will fail until JSON-B feature is released this Friday.

sishida commented on README.adoc:
```diff
@@ -112,6 +111,19 @@ include::finish/src/main/java/io/openliberty/guides/consumingrest/model/Artist.j
+JSON-B is built on top of the existing JSON-P API. JSON-B can do everything that JSON-P can do
+and allows for more customization for serializing and deserializing. For
+more information on JSON-B and how to serialize POJOs into JSON strings please refer to the
+http://json-b.net/users-guide.html[JSON Binding 1.0 Users Guide].
```
Final comments from ID:
- Line 116: We try to avoid bold (and italics) for emphasis. It draws unnecessary attention to the word. For example, we wouldn't want an inexperienced user to mistake the bold word for a term.
- Line 124: Comma needed after "strings."
- Line 124: We try to omit "please" when giving instructions. If we added it in front of every imperative statement like this one, it could quickly get out of hand.
- If you think that users would be unfamiliar with any of the acronyms, in the first instance of the acronym, spell out the words, such as "JavaScript Object Notation (JSON)." All subsequent instances can then just be "JSON."

sishida commented:
@KyleAure: Thanks for putting up with all my updates! ;) I have added a last comment for the new section. Thank you!

Kubik42 commented:
Merged into diff branch temporarily to see how everything looks on the qa site

yeekangc commented:
@Kubik42 @evelinec, we need to address the below before we merge into master and publish to the live site:

- Explore how to access a simple RESTful web service and consume its resources in Java _using JSON-B and JSON-P_.

- Code formatting: Extra spaces or tabs for some lines in classes that are included in the guide.

- "By default, JSON-B requires a public default no-argument constructor for deserialization": Requires a default no-argument constructor where?

- "Note that the Artist POJO, does not require this annotation since all instance members match the JSON keys by name": "Note that" is usually unnecessary per @sishida. Comma splice too? _The Artist POJO does not require this annotation since all instance members match the JSON keys by name._

- "However, if you are consuming a REST resource from a third party service, you can separate your GET and PULL requests from your data consumption": PULL requests?

- "You can find more detailed explanations on how to define your own system properties and how to access them via the System.getProperty call in the MicroProfile guide": We have archived the MicroProfile guide so this needs to be updated or dropped.

Please address the above.  Then, we can proceed with merging, final testing and publishing.

sishida commented:
Yes, I'd say, "The `Artist` POJO does not require this annotation because all instance members match the JSON keys by name." :)

KyleAure commented:
Good morning @yeekangc and @sishida,

- [x] MADE THIS CHANGE Explore how to access a simple RESTful web service and consume its resources in Java using JSON-B and JSON-P.

- [ ] Code formatting: Extra spaces or tabs for some lines in classes that are included in the guide. Some lines of code don't align properly.

Can you give examples of where you are seeing extra spaces/tabs.  

- [ ] "By default, JSON-B requires a public default no-argument constructor for deserialization": Requires a default no-argument constructor where?

I made comments on the constructors for both POJO's below.  If you think more explanation needs to be put in let me know.  Default no-argument constructors should be well known by any developer that is reading this guide.
From Artist class: `public Artist() { //default constructor can be defined }`
From Album class: `public Album() { //default constructor can be defined }`

- [x] USED @sishida's SUGGESTION "Note that the Artist POJO, does not require this annotation since all instance members match the JSON keys by name": "Note that" is usually unnecessary per @sishida. Comma splice too? Consider The Artist POJO does not require this annotation since all instance members match the JSON keys by name.

- [x] CHANED TO POST "However, if you are consuming a REST resource from a third party service, you can separate your GET and PULL requests from your data consumption": PULL requests?

- [x] REMOVED "You can find more detailed explanations on how to define your own system properties and how to access them via the System.getProperty call in the MicroProfile guide": We have archived the MicroProfile guide so this needs to be updated or dropped.

yeekangc commented:
@KyleAure:

- Code formatting: Please see https://qa-guides.mybluemix.net/guides/rest-client-java.html.  For example, the lines are not properly aligned in the Artist and Album classes shown.  Just look through all the classes that are being included.

- Yes, folks will know what's default no-arg constructor and I know what we are trying to say here.  It will be clearer if we will indicate that this is meant for the POJO that you will annotate and eventually bind.  That's it.  =)

Thanks!
